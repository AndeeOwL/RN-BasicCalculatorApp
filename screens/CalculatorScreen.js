import { View, StyleSheet } from "react-native";
import { useState, useCallback } from "react";

import Buttons from "../components/Buttons";
import Operators from "../components/Operators";
import { StatusBar } from "expo-status-bar";
import { buttonPressHandler } from "../services/buttonPressHandler";
import { operationPressHandler } from "../services/operationPressHandler";
import ResultField from "../components/ResultField";
import CalculationField from "../components/CalculationField";

function CalculatorScreen() {
  const [resultText, setResultText] = useState("");
  const [calculationText, setCalculationText] = useState("0");
  const lastChar = useCallback(
    () => calculationText.charAt(calculationText.length - 1),
    [calculationText]
  );
  let result = [];

  const onButtonPress = (enteredValue) => {
    result = buttonPressHandler(enteredValue, calculationText, lastChar);
    console.log(result[0]);
    console.log(result[1]);
    if (result.length === 1) {
      setCalculationText(result[0]);
    } else {
      setCalculationText(result[1]);
      setResultText(result[0]);
    }
  };

  const onOperationPress = (enteredValue) => {
    result = operationPressHandler(calculationText, enteredValue, lastChar);

    if (result.length === 1) {
      setCalculationText(result[0]);
    } else {
      setResultText(result[1]);
      setCalculationText(result[0]);
    }
  };

  return (
    <View style={styles.rootContainer}>
      <StatusBar style='light' />
      <ResultField value={resultText} />
      <CalculationField value={calculationText} />
      <View style={styles.buttonsContainer}>
        <Buttons
          style={styles.buttonsContainer}
          onButtonPress={onButtonPress}
          onOperationPress={onOperationPress}
        />
        <Operators
          onButtonPress={onButtonPress}
          onOperationPress={onOperationPress}
        />
      </View>
    </View>
  );
}

export default CalculatorScreen;

const styles = StyleSheet.create({
  rootContainer: {
    flex: 1,
    backgroundColor: "black",
  },
  buttonsContainer: {
    flex: 6,
    flexDirection: "row",
    backgroundColor: "orange",
  },
});
