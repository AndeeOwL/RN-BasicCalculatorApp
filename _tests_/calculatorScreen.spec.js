import { render, fireEvent } from "@testing-library/react-native";
import CalculatorScreen from "../screens/CalculatorScreen";

describe("Tests calculatorScreen component", () => {
  test("renders correctly", () => {
    const tree = render(<CalculatorScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("should change values to calculationField and resultField on = press", () => {
    const { getByTestId, getByText } = render(<CalculatorScreen />);
    const resultField = getByTestId("resultField");
    const calculationField = getByTestId("calculationField");

    fireEvent.press(getByText("1"));
    fireEvent.press(getByText("+"));
    fireEvent.press(getByText("3"));
    fireEvent.press(getByText("="));
    expect(calculationField.children).toEqual(["4"]);
    expect(resultField.children).toEqual(["4"]);
  });
  test("should change values to calculationField on DEL press", () => {
    const { getByTestId, getByText } = render(<CalculatorScreen />);
    const resultField = getByTestId("resultField");
    const calculationField = getByTestId("calculationField");

    fireEvent.press(getByText("5"));
    fireEvent.press(getByText("-"));
    fireEvent.press(getByText("5"));
    fireEvent.press(getByText("DEL"));
    expect(calculationField.children).toEqual(["5-"]);
  });
  test("should change values to calculationField on AC press", () => {
    const { getByTestId, getByText } = render(<CalculatorScreen />);
    const resultField = getByTestId("resultField");
    const calculationField = getByTestId("calculationField");

    fireEvent.press(getByText("9"));
    fireEvent.press(getByText("*"));
    fireEvent.press(getByText("9"));
    fireEvent.press(getByText("AC"));
    expect(calculationField.children).toEqual(["0"]);
    expect(resultField.children).toEqual(["0"]);
  });
  test("should change values to calculationField on AC press", () => {
    const { getByTestId, getByText } = render(<CalculatorScreen />);
    const resultField = getByTestId("resultField");
    const calculationField = getByTestId("calculationField");

    fireEvent.press(getByText("2"));
    fireEvent.press(getByText("*"));
    fireEvent.press(getByText("2"));
    fireEvent.press(getByText("+"));
    expect(calculationField.children).toEqual(["4+"]);
    expect(resultField.children).toEqual(["4"]);
  });
  test("should change values to calculationField on AC press", () => {
    const { getByTestId, getByText } = render(<CalculatorScreen />);
    const resultField = getByTestId("resultField");
    const calculationField = getByTestId("calculationField");

    fireEvent.press(getByText("2"));
    fireEvent.press(getByText("+"));
    fireEvent.press(getByText("+"));
    expect(calculationField.children).toEqual(["2+"]);
  });
});
