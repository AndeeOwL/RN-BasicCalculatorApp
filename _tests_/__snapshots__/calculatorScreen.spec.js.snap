// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests calculatorScreen component renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "borderBottomColor": "Orange",
        "borderBottomWidth": 4,
        "flex": 4,
        "justifyContent": "flex-end",
        "marginRight": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 50,
          "fontWeight": "bold",
          "padding": 10,
        }
      }
      testID="resultField"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "grey",
        "borderColor": "black",
        "borderWidth": 3,
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 30,
        }
      }
      testID="calculationField"
    >
      0
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "flex": 6,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderColor": "black",
            "borderWidth": 3,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            1
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            2
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            3
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderColor": "black",
            "borderWidth": 3,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            4
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            5
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            6
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderColor": "black",
            "borderWidth": 3,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            7
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            8
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            9
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderColor": "black",
            "borderWidth": 3,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            .
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            0
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={false}
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            =
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "grey",
          "borderColor": "black",
          "borderWidth": 3,
          "flex": 1,
          "justifyContent": "space-around",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 30,
              "fontWeight": "bold",
            }
          }
        >
          DEL
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 30,
              "fontWeight": "bold",
            }
          }
        >
          AC
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 30,
              "fontWeight": "bold",
            }
          }
        >
          -
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 30,
              "fontWeight": "bold",
            }
          }
        >
          +
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 30,
              "fontWeight": "bold",
            }
          }
        >
          /
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 30,
              "fontWeight": "bold",
            }
          }
        >
          *
        </Text>
      </View>
    </View>
  </View>
</View>
`;
